### DISCUSSION TITLE
Using zrepl in conjuction with zsys

### DISCUSSION THREAD

**Discussion by darkbasic** [zrepl/zrepl#642] [Q&A]
In https://github.com/zrepl/zrepl/issues/494 I said I'd like to be able to clone and promote a rootfs snapshot at boot time -directly from grub- because the root cannot be easily unmounted at will (also you couldn't be able to boot at all if an update broke your system). Until zrepl gains that functionality I would like to mix it with zsys, with the latter being in charge of snapshotting bootfs, rootfs and home and the former the other datasets plus replication.
I've read the documentation and there are no examples of how to achieve this. I guess I will have to exclude bootfs, roots and home from the main job and create a new one with `manual` snapshotting type but apart from that I'm completely blind.

Will zrepl still take care of bookmarks creation? zsys could delete these snapshots at any point...
Will I have to run `zrepl signal wakeup JOB` from the zsys post snapshot hook? (I'm not even sure if zsys allows you to run custom scripts after the snapshot creation)
Will reverting the rootfs/boots/home dataset break replication? (AFAIK zsys uses clones so it shouldn't)

Can you please help me to set this up?

**Comment by problame** [comment]
I don't know enough about how zsys manages datasets.

In general, you can follow the https://zrepl.github.io/quickstart/backup_to_external_disk.html quickstart guide as an example for decoupling snapshot management from replication.

I want to revisit zrepl snapshot management in the coming months.

---

> Will zrepl still take care of bookmarks creation? zsys could delete these snapshots at any point...

The replication cursor bookmark is managed by the replication job.
As long as zsys doesn't delete it, everything should be fine.

>  Will I have to run zrepl signal wakeup JOB from the zsys post snapshot hook? (I'm not even sure if zsys allows you to run custom scripts after the snapshot creation)

Yes. Or you could just do the wake-up from `cron` periodically. I plan to integrate periodic replication triggers (decoupled from snapshot creation) in the coming months.

>  Will reverting the rootfs/boots/home dataset break replication? (AFAIK zsys uses clones so it shouldn't)

If it uses clones, things *could* work out. The problem is that zrepl doesn't mirror clone relationship right now, i.e., a new clone will be fully replicated, even if the clone's origin snapshot is present on the receiver.
At this point, I'm embarrassed to point to the original issue, #71 . It's kind of blocked by #408 , which isn't seeing progress because I don't have enough time right now.

**Reply by darkbasic** [reply]
> The problem is that zrepl doesn't mirror clone relationship right now, i.e., a new clone will be fully replicated, even if the clone's origin snapshot is present on the receiver.

Oh, that's a **big** showstopper unfortunately.

Bandwidth wouldn't be a problem (I have a dedicated 10Gbps DAC for replication) but it will take way too much space on the remote NAS.
The root fs won't be too big compared to the other datasets and I (hopefully) won't have to revert it too frequently so it could still be a manageable flaw. What isn't manageable is restoring the **other** datasets.

This is a virtualization server backed with expensive (and very small) Optane storage so I don't want to keep the snapshots around, I want to replicate them to the NAS with mechanical hard drives. Each VM has its own zvol dataset. Unfortunately it's not clear what's the most suitable way of restoring for the zrepl workflow and there is no documentation at all (which is a pity, because a backup solution is only as good as your ability to restore it). I thought clones would be the best choice because they allow you to retain your whole history and don't break incremental snapshots, but I guess that would not be feasible because of #71. Contrary to the root fs I will create vm clones very often and these datasets are much larger. I guess there must be a better way to restore with zrepl, but I can't figure it out on my own because of the lack of documentation. Can you please help me?

**Reply by problame** [reply]
Sorry for the late reply.

As per restore: yes, zrepl users still need to know a lot of ZFS right now.

Generally, there are three kinds of  restore with any ZFS-based backup solution:

1. primary completely lost, restore some version (with our without history) from the backup server
2. primary still there, but need to rollback to an earlier snapshot that is still present on the primary
  - case 1: earlier snapshot has already been seen & started / finished replicating by zrepl
  - case 2: earlier snapshot has not yet been seen / replicated by zrepl
3. primary still there, but need to rollback to a snapshot that is only present on the backup
  - case 1: there is a common ancestor between primary and backup to which we can rollback the primary, then incremental-recv up to the restore snapshot from the backup
  - case 2: there is no common ancestor between primary and backup

---

How to do it:

(1) is easy. Use `zfs send/recv` from backup to primary.

(2) case 2: stop zrepl, do zfs rollback, start zrepl. Done.

(2) case 1 and (3) case 1: on primary, rollback to the the target snapshot. That's it for (2). For (3), then do an `zfs send/recv` from backup to primary to close the gap. 

(3) case 2: destroy/rename-away the dataset on the primary, then do full `zfs send/recv` from backup to primary. 

---

Clones: since zrepl doesn't supporot replicating clones as clones (they'll be full datasets on the receiver, with just the clone-specific snapshots present), your only hope to restore them _as clones_ is to ensure that the clone's `origin` snapshot is present on the backup server. During restore, you can then use `zfs recv -o origin=...`. I won't go into the details there, as zrepl doesn't support clones.

---

Hope this helps.

**Reply by darkbasic** [reply]
> (2) case 1 and (3) case 1: on primary, rollback to the the target snapshot

@problame rolling back on `primary` removes all subsequent snapshots on `primary`. What will happen on `backup`? Won't that trigger removal of all subsequent snapshots on `backup` as well? I will basically lose any access to all snapshots after the one to which I rolled back to.

**Reply by problame** [reply]
The current version of zrepl never destroys receive-side state. You will get a replication conflict after the rollback, complaining that the receiver has newer versions than the sender.

**Reply by darkbasic** [reply]
Ok, thanks for the detailed explanation. I really hope that zrepl will support clones sooner rather than later because that will make restoring so much more convenient.
